File: TMessagesProj/src/main/java/org/telegram/messenger/NotificationCenter.java
/*
 * This is the source code of Telegram for Android v. 11.x.x.
 * It is licensed under GNU GPL v. 2 or later.
 * You should have received a copy of the license in this archive (see LICENSE).
 *
 * Copyright Nikolai Kudashov, 2013-2024.
 */

package org.telegram.messenger;

import java.util.ArrayList;
import java.util.HashMap;

public class NotificationCenter {
    private static int totalEvents = 1;

    public static final int didReceiveNewMessages = totalEvents++;
    public static final int updateInterfaces = totalEvents++;
    public static final int dialogsNeedReload = totalEvents++;
    public static final int closeChats = totalEvents++;
    public static final int messagesDeleted = totalEvents++;
    public static final int messagesRead = totalEvents++;
    public static final int messagesReadContent = totalEvents++;
    public static final int messageReceivedByAck = totalEvents++;
    public static final int messageSendError = totalEvents++;
    public static final int userStatusUpdated = totalEvents++;
    public static final int didUpdateConnectionState = totalEvents++;
    public static final int activeUserUpdated = totalEvents++;
    public static final int encryptedChatCreated = totalEvents++;
    public static final int chatInfoDidLoad = totalEvents++;
    public static final int channelRightsUpdated = totalEvents++;
    public static final int userFullInfoDidLoad = totalEvents++;
    public static final int scheduledMessagesDidLoad = totalEvents++;
    public static final int topicsDidLoaded = totalEvents++;
    public static final int notificationsSettingsUpdated = totalEvents++;
    public static final int reactionsDidLoad = totalEvents++;
    public static final int webPageDidLoad = totalEvents++;
    public static final int botAppDidLoad = totalEvents++;
    public static final int groupCallUpdated = totalEvents++;
    public static final int messagesDidLoaded = totalEvents++;

    public static final int UPDATE_MASK_READ_DIALOG_MESSAGE = 0x0001;
    public static final int UPDATE_MASK_SELECT_DIALOG = 0x0002;
    public static final int UPDATE_MASK_SEND_STATE = 0x0004;

    public interface NotificationCenterDelegate {
        void didReceivedNotification(int id, int account, Object... args);
    }

    private static volatile NotificationCenter[] Instance = new NotificationCenter[AccountInstance.MAX_ACCOUNT_COUNT];
    private final int currentAccount;
    private final HashMap<Integer, ArrayList<NotificationCenterDelegate>> observers = new HashMap<>();

    public static NotificationCenter getInstance(int num) {
        NotificationCenter localInstance = Instance[num];
        if (localInstance == null) {
            synchronized (NotificationCenter.class) {
                localInstance = Instance[num];
                if (localInstance == null) {
                    Instance[num] = localInstance = new NotificationCenter(num);
                }
            }
        }
        return localInstance;
    }

    public NotificationCenter(int instance) {
        currentAccount = instance;
    }

    public void addObserver(NotificationCenterDelegate observer, int id) {
        ArrayList<NotificationCenterDelegate> arrayList = observers.get(id);
        if (arrayList == null) {
            arrayList = new ArrayList<>();
            observers.put(id, arrayList);
        }
        if (!arrayList.contains(observer)) {
            arrayList.add(observer);
        }
    }

    public void removeObserver(NotificationCenterDelegate observer, int id) {
        ArrayList<NotificationCenterDelegate> arrayList = observers.get(id);
        if (arrayList != null) {
            arrayList.remove(observer);
        }
    }

    public void postNotificationName(int id, Object... args) {
        ArrayList<NotificationCenterDelegate> arrayList = observers.get(id);
        if (arrayList != null && !arrayList.isEmpty()) {
            ArrayList<NotificationCenterDelegate> arrayListCopy = new ArrayList<>(arrayList);
            for (NotificationCenterDelegate delegate : arrayListCopy) {
                delegate.didReceivedNotification(id, currentAccount, args);
            }
        }
    }
}

File: src/core/NotificationCenter.ts
/**
 * NotificationCenter.ts - Telegram Central Event Bus
 * 
 * Replicated directly from DrKLO/Telegram Android:
 * org.telegram.messenger.NotificationCenter.java
 */

export interface NotificationCenterDelegate {
  didReceivedNotification(id: number | string, account: number, ...args: any[]): void;
}

export class NotificationCenter {
  // DrKLO/Telegram Android Event Constants
  public static readonly didReceiveNewMessages = 1;
  public static readonly updateInterfaces = 2;
  public static readonly dialogsNeedReload = 3;
  public static readonly closeChats = 4;
  public static readonly messagesDidLoad = 5;
  public static readonly didReceivedWebsitesList = 6;
  public static readonly didReplacedPhotoInMemCache = 7;
  public static readonly notificationsCountUpdated = 8;
  public static readonly didUpdateConnectionState = 9;
  public static readonly userFullInfoDidLoad = 10;
  public static readonly pinnedInfoDidLoad = 11;
  public static readonly messagePlayingProgressDidChanged = 12;
  public static readonly messagePlayingDidReset = 13;
  public static readonly messagePlayingPlayStateChanged = 14;
  public static readonly recordProgressChanged = 15;
  public static readonly recordStartError = 16;
  public static readonly recordStopped = 17;
  public static readonly chatDidCreated = 18;
  public static readonly chatDidFailCreate = 19;
  public static readonly chatInfoDidLoad = 20;
  public static readonly contactsDidLoad = 21;
  public static readonly userSelectedEmoji = 22;
  public static readonly userSelectedSticker = 23;
  public static readonly themeDidLoad = 24;
  public static readonly needSetDayNightTheme = 25;
  public static readonly didReceivedDraft = 26;
  public static readonly messageReceivedByAck = 27;
  public static readonly messagesDeleted = 28;
  public static readonly messagesRead = 29;
  public static readonly didClearDatabase = 30;
  public static readonly checkClientRole = 31;
  public static readonly storiesUpdated = 32;
  public static readonly topicsDidLoaded = 33;
  public static readonly privacyRulesUpdated = 34;
  public static readonly mainUserInfoChanged = 35;
  public static readonly UPDATE_MASK_READ_DIALOG_MESSAGE = 0x0001;
  public static readonly UPDATE_MASK_SELECT_DIALOG = 0x0002;
  public static readonly UPDATE_MASK_SEND_STATE = 0x0004;

  private static instances = new Map<number, NotificationCenter>();
  private static globalInstance: NotificationCenter;

  private observers = new Map<number | string, Set<NotificationCenterDelegate | ((...args: any[]) => void)>>();
  private currentAccount: number;

  public static getInstance(account: number = 0): NotificationCenter {
    if (!NotificationCenter.instances.has(account)) {
      const inst = new NotificationCenter(account);
      NotificationCenter.instances.set(account, inst);
      if (account === 0 && !NotificationCenter.globalInstance) {
        NotificationCenter.globalInstance = inst;
      }
    }
    return NotificationCenter.instances.get(account)!;
  }

  public static getGlobalInstance(): NotificationCenter {
    if (!NotificationCenter.globalInstance) {
      NotificationCenter.globalInstance = NotificationCenter.getInstance(0);
    }
    return NotificationCenter.globalInstance;
  }

  private constructor(account: number = 0) {
    this.currentAccount = account;
  }

  public addObserver(observer: NotificationCenterDelegate | ((...args: any[]) => void), id: number | string): void {
    if (!this.observers.has(id)) {
      this.observers.set(id, new Set());
    }
    this.observers.get(id)!.add(observer);
  }

  public removeObserver(observer: NotificationCenterDelegate | ((...args: any[]) => void), id: number | string): void {
    const list = this.observers.get(id);
    if (list) {
      list.delete(observer);
      if (list.size === 0) {
        this.observers.delete(id);
      }
    }
  }

  public postNotificationName(id: number | string, ...args: any[]): void {
    const list = this.observers.get(id);
    if (list) {
      list.forEach((obs) => {
        try {
          if (typeof obs === 'function') {
            obs(...args);
          } else if (typeof obs.didReceivedNotification === 'function') {
            obs.didReceivedNotification(id, this.currentAccount, ...args);
          }
        } catch (e) {
          console.error('[NotificationCenter] Error in observer callback:', e);
        }
      });
    }

    // Also dispatch a browser CustomEvent for reactive DOM integration
    if (typeof window !== 'undefined') {
      window.dispatchEvent(
        new CustomEvent('tg-notification-center', {
          detail: { id, account: this.currentAccount, args },
        })
      );
    }
  }

  public hasObservers(id: number | string): boolean {
    return (this.observers.get(id)?.size || 0) > 0;
  }
}

export const notificationCenter = NotificationCenter.getInstance(0);
